### PR TITLE
Cart functionality

### PR DESCRIPTION
**Pull request by pankaj-prog**
Changes done : 
- cart context to maintain cart products state
- update navbar with cart quantity badge
- update course card to display rating 
- course details page CTA - add to cart 
- cart bill calculation based on cart products

### CODE REVIEW

File: src/pages/Cart/components/CartBill/cartBillUtils.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+export const cartInitialPrice = (cartProducts) => {
```
Comment (m26unkwn) [inline src/pages/Cart/components/CartBill/cartBillUtils.js:1]: You could have done it in one function but anyhow this is quite readable.

**Review by m26unkwn — APPROVED** [review]
Your code is clean and readable.

Things you can do on the UI part [edge cases]
1. While updating the cart user is able to decrease the quantity below the zero
2. While clicking on Addtocart Button if you are clicking it continuously its updates the cart with respect to clicks.

Otherwise everything is good to go.

File: src/components/Navbar/Navbar.jsx [hunk 2]
```diff
@@ -57,8 +58,12 @@ const Navbar = () => {
...
             <i className="far fa-heart"></i>
           </NavLink>
-          <NavLink to="/cart" className="btn icon-btn link">
-            <i className="fas fa-shopping-cart"></i>
+          <NavLink
+            to="/cart"
+            className="btn icon-btn link badge-container icon-container"
+          >
+            <i className="fas fa-shopping-cart "></i>
+            {encodedToken && <span class="badge">{cartProducts.length}</span>}
```
Comment (PuneetP16) [inline src/components/Navbar/Navbar.jsx:66]: Suggestion: You could use Navlink className isActive function to style the currently selected page.

For reference check: https://ultimatecourses.com/blog/active-navlink-classes-with-react-router

File: src/context/utils/cartUtils.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        "error from product quantity handler with status",
+        response.status
+      );
+    }
+  } catch (error) {
+    console.log(error);
+  }
+};
+
+export const cartHandlers = {
```
Comment (PuneetP16) [inline src/context/utils/cartUtils.js]: Addon Suggestion: you could show error message when user enters wrong email or password by catching the errors.

image

File: src/pages/Cart/Cart.jsx [hunk 4]
```diff
@@ -2,73 +2,60 @@ import React from "react";
...
+              <CartBill />
             </div>
-            <div className="detail fw-b">
-              Final Price: <span className="cart-total-price">₹4919</span>
+          ) : (
+            <div className="text-center">
+              <h3>Your cart is currently empty 😢</h3>
+              <p className="gutter-bottom-8">
+                Before proceeding to checkout, you must add some products to
+                your cart.{" "}
```
Comment (PuneetP16) [inline src/pages/Cart/Cart.jsx:36]: Clear empty Jsx strings.

**Review by PuneetP16 — APPROVED** [review]
Your code is super clean, liked your way of writing code. Good to merge.
Gave an Addon Suggestion just to improve Users Experience.

File: src/components/Navbar/Navbar.jsx (hunk 2 above)
Comment (pankaj-prog) [reply src/components/Navbar/Navbar.jsx:66]: Thank you Puneet

**Pull request merged**